### PR TITLE
Fix issue with plz test not printing panic output

### LINKED ISSUE #2752: `plz test` does not print test output when there's any logs to stderr
Example:
```
func foo() {
	log.Error("hello world")
	panic("goodbye world")
}

func TestFoo(t *testing.T) {
	foo()
}
```

Expected behaviour:
* Test fails
* log is printed
* panic is printed with trace
* it is clear that the panic comes from TestFoo

Actual behaviour:
* Test fails
* log is printed

Removing the log line causes the panic to be printed correctly,

### PR DESCRIPTION
**Pull request by samwestmoreland**
For reasons I don't fully understand, the panic output gets sent through stdout. Currently we don't collect the stdout from the exec call in `runTest`.

This change gets both stdout and stderr from the exec, and also tries to straighten out the capturing of stdout and stderr in `src/process/process.go`, which feels like it could have unforeseen consequences so definitely less sure about that bit.

Before:
```
Fail: //test/misc_rules:go_test   0 passed   0 skipped   1 failed   0 errored Took 10ms
Failure:  in TestFoo
2023/04/20 17:15:13 hello world
Standard error:
2023/04/20 17:15:13 hello world

//test/misc_rules:go_test 1 test run in 8ms; 0 passed, 1 failed
4 test targets and 4 tests run; 3 passed, 1 failed.
Total time: 70ms real, 30ms compute.
Messages:
17:15:13.826   ERROR: //test/misc_rules:go_test failed
```
After:
```
Fail: //test/misc_rules:go_test   0 passed   0 skipped   1 failed   1 errored Took 10ms                                         
Failure:  in TestFoo   
2023/04/20 17:16:09 hello world
Standard error:
2023/04/20 17:16:09 hello world

Error: ReturnValue in go_test
Test returned 0 but still reported failures

Standard output:
=== RUN   TestFoo
2023/04/20 17:16:09 hello world
--- FAIL: TestFoo (0.00s)
panic: goodbye world [recovered]
        panic: goodbye world

goroutine 6 [running]:
testing.tRunner.func1.2({0x512f20, 0x565c30})
        /home/swestmoreland/please_dev/please/plz-out/bin/third_party/go/toolchain/src/testing/testing.go:1396 +0x24e
testing.tRunner.func1()
        /home/swestmoreland/please_dev/please/plz-out/bin/third_party/go/toolchain/src/testing/testing.go:1399 +0x39f
panic({0x512f20, 0x565c30})
        /home/swestmoreland/please_dev/please/plz-out/bin/third_party/go/toolchain/src/runtime/panic.go:884 +0x212
go_test_lib.foo(...)
        test/misc_rules/test.go:12
go_test_lib.TestFoo(0x0?)
        test/misc_rules/test.go:16 +0x4d                                                                
testing.tRunner(0xc000118820, 0x542400)                                                           
        /home/swestmoreland/please_dev/please/plz-out/bin/third_party/go/toolchain/src/testing/testing.go:1446 +0x10b
created by testing.(*T).Run                  
        /home/swestmoreland/please_dev/please/plz-out/bin/third_party/go/toolchain/src/testing/testing.go:1493 +0x35f
                                     
//test/misc_rules:go_test 2 tests run in 7ms; 0 passed, 1 errored, 1 failed
4 test targets and 5 tests run; 3 passed, 1 errored, 1 failed.
Total time: 50ms real, 20ms compute.
Messages:
17:16:09.710   ERROR: //test/misc_rules:go_test failed
```
Fixes #2752

### CODE REVIEW

**Comment by samwestmoreland** [comment]
Ok there's a failing test, which is handy. Will try and figure this out again tomorrow. Looks like the panic does come through stderr when there's no log.Error before it, but when there is a log.Error the panic gets redirected through stdout.

File: src/test/test_step.go [hunk 1]
```diff
@@ -347,14 +347,14 @@ func testCommandAndEnv(state *core.BuildState, target *core.BuildTarget, run int
...
 
-func runTest(state *core.BuildState, target *core.BuildTarget, run int) ([]byte, error) {
+func runTest(state *core.BuildState, target *core.BuildTarget, run int) ([]byte, []byte, error) {
 	replacedCmd, env, err := testCommandAndEnv(state, target, run)
 	if err != nil {
-		return nil, err
+		return nil, nil, err
 	}
 	log.Debugf("Running test %s#%d\nENVIRONMENT:\n%s\n%s", target.Label, run, strings.Join(env, "\n"), replacedCmd)
-	_, stderr, err := state.ProcessExecutor.ExecWithTimeoutShellStdStreams(target, target.TestDir(run), env, target.Test.Timeout, state.ShowAllOutput, false, process.NewSandboxConfig(target.Test.Sandbox, target.Test.Sandbox), replacedCmd, state.DebugFailingTests)
```
Comment (Tatskaari) [inline src/test/test_step.go:356]: I think `stderr` is badly named here. It looks like it should contain all the output from both stdout and stderr, so this should've worked fine.

File: src/test/test_step.go [hunk 2]
```diff
@@ -384,8 +384,8 @@ func doTestResults(tid int, state *core.BuildState, target *core.BuildTarget, ru
 		}
 	} else {
 		var stdout []byte
-		stdout, err = prepareAndRunTest(tid, state, target, run)
```
Comment (Tatskaari) [inline src/test/test_step.go:387]: Oooff and now the value of the variable called `stderr` from above is stored in a variable called `stdout`... This is very confusing.

**Review by Tatskaari — CHANGES_REQUESTED** [review]
I don't know if this is quite right; we don't want please printing extra stuff to stdout. I'm not sure I see why we're not seeing the output from the panic, though? We definitely do report both stdout and stderr in test results under normal operation. Maybe we can have a look at this together.

**Comment by samwestmoreland** [comment]
After the latest change:
```
Fail: //experimental/swestmoreland/plz_test:go_test   0 passed   0 skipped   1 failed   0 errored Took 10ms                                      
Failure:  in TestFoo                                                                                                                             
time="2023-04-24T15:48:59+01:00" level=error msg="hello world" file="experimental/swestmoreland/plz_test/test.go:11" func= kind=application      
panic: goodbye world [recovered]                                                                                                                 
        panic: goodbye world                                                                                                                     
                                                                                                                                                 
goroutine 6 [running]:                                                                                                                           
testing.tRunner.func1.2({0x221820, 0x2da238})                                                                                                    
        /opt/tm/tools/go/1.19.8/usr/go/src/testing/testing.go:1396 +0x24e                                                                        
testing.tRunner.func1()                                                                                                                          
        /opt/tm/tools/go/1.19.8/usr/go/src/testing/testing.go:1399 +0x39f                                                                        
panic({0x221820, 0x2da238})                                                                                                                      
        /opt/tm/tools/go/1.19.8/usr/go/src/runtime/panic.go:884 +0x212                                                                           
go_test_lib.foo(...)                                                                                                                             
        experimental/swestmoreland/plz_test/test.go:12                                                                                           
go_test_lib.TestFoo(0x0?)                                                                                                                        
        experimental/swestmoreland/plz_test/test.go:16 +0x65                                                                                     
testing.tRunner(0xc000148b60, 0x28da48)                                                                                                          
        /opt/tm/tools/go/1.19.8/usr/go/src/testing/testing.go:1446 +0x10b                                                                        
created by testing.(*T).Run                                                                                                                      
        /opt/tm/tools/go/1.19.8/usr/go/src/testing/testing.go:1493 +0x35f                                                                        
                                                                                                                                                 
Standard error:                                                                                                                                  
time="2023-04-24T15:48:59+01:00" level=error msg="hello world" file="experimental/swestmoreland/plz_test/test.go:11" func= kind=application      
                                                                                                                                                 
//experimental/swestmoreland/plz_test:go_test 1 test run in 9ms; 0 passed, 1 failed                                                              
    TestFoo  FAIL  0s                                                                                                                            
1 test target and 1 test run; 0 passed, 1 failed.                                                                                                
Total time: 180ms real, 10ms compute.                                                                                                            
Messages:                                                                                                                                        
15:48:59.358   ERROR: //experimental/swestmoreland/plz_test:go_test failed                                                                       
```

**Review by goddenrich — APPROVED** [review]
Looks good to me

**Pull request merged**